tools/cassandra_delete_range/main.go

```diff
@@ -0,0 +1,544 @@
+	var info = deleteInfo{Query: deleteQueryTemplate}
+
+	go func() {
+		for params := range outChannel {
+			info.Data = append(info.Data, params)
+		}
+	}()
+
+	var wg sync.WaitGroup
+	var sessionCreationWaitGroup sync.WaitGroup
+	var totalRows uint64
+	var totalErrors uint64
+
+	wg.Add(numberOfParallelClientThreads)
+	sessionCreationWaitGroup.Add(numberOfParallelClientThreads)
+
+	for i := 0; i < numberOfParallelClientThreads; i++ {
+		go func() {
+			defer wg.Done()
+
+			var session *gocql.Session
+			var err error
+			if session, err = cluster.CreateSession(); err == nil {
+				defer session.Close()
+
+				sessionCreationWaitGroup.Done()
+				sessionCreationWaitGroup.Wait()
+				preparedQueryString := fmt.Sprintf(queryTemplate, "?", "?")
+				preparedQuery := session.Query(preparedQueryString)
+
+				for r := range rangesChannel {
+					preparedQuery.Bind(r.StartRange, r.EndRange)
+					iter := preparedQuery.Iter()
+
+					var key []byte
+					var seq uint64
+					var rowsRetrieved uint64
+
+					for iter.Scan(&key, &seq) {
+						rowsRetrieved++
+
+						// only grab the rows that are in the correct range of sequence numbers
+						if fromLedgerIdx <= seq {
+							outChannel <- deleteParams{Seq: seq, Blob: key}
+						}
+					}
+
+					if err := iter.Close(); err != nil {
+						log.Printf("ERROR: iteration failed: %s\n", err)
+						atomic.AddUint64(&totalErrors, 1)
+					}
+
+					atomic.AddUint64(&totalRows, rowsRetrieved)
+				}
+			} else {
+				log.Printf("ERROR: %s\n", err)
+				atomic.AddUint64(&totalErrors, 1)
+			}
+		}()
+	}
+
+	wg.Wait()
+	close(outChannel)
+
+	return info, totalRows, totalErrors
+}
+
+func performDeleteQueries(cluster *gocql.ClusterConfig, info *deleteInfo, colSettings columnSettings) (uint64, uint64) {
+	var wg sync.WaitGroup
+	var sessionCreationWaitGroup sync.WaitGroup
+	var totalDeletes uint64
+	var totalErrors uint64
+
+	chunks := splitDeleteWork(info)
+	chunksChannel := make(chan []deleteParams, len(chunks))
+	for i := range chunks {
+		chunksChannel <- chunks[i]
+	}
+
+	close(chunksChannel)
+
+	wg.Add(len(chunks))
+	sessionCreationWaitGroup.Add(len(chunks))
+
+	for i := 0; i < len(chunks); i++ {
+		go func() {
+			defer wg.Done()
+
+			var session *gocql.Session
+			var err error
+			if session, err = cluster.CreateSession(); err == nil {
+				defer session.Close()
+
+				sessionCreationWaitGroup.Done()
+				sessionCreationWaitGroup.Wait()
+				preparedQuery := session.Query(info.Query)
+
+				var bindCount = strings.Count(info.Query, "?")
+
+				for chunk := range chunksChannel {
+					for _, r := range chunk {
+						if bindCount == 2 {
+							preparedQuery.Bind(r.Blob, r.Seq)
+						} else if bindCount == 1 {
+							if colSettings.UseSeq {
+								preparedQuery.Bind(r.Seq)
+							} else if colSettings.UseBlob {
+								preparedQuery.Bind(r.Blob)
+							}
+						}
+
+						if err := preparedQuery.Exec(); err != nil {
+							log.Printf("DELETE ERROR: %s\n", err)
```

Comment:
Could you please add more details to the log about error, for example failed query?

And also I think we should log errors to stderr.